File: workflows/subtasks.py
"""
Subtask execution examples.

Demonstrates how tasks can call other tasks using await syntax.
"""

import asyncio
import logging
from app import app
from basic_tasks import square, multiply

logger = logging.getLogger(__name__)

@app.task
async def add_squares(a: int, b: int) -> int:
    """
    Add the squares of two numbers by calling the square task twice.

    This demonstrates subtask execution pattern.
    """
    logger.info(f"Computing add_squares: {a}² + {b}²")

    result1, result2 = await asyncio.gather(square(a), square(b))
    logger.info(f"First square result: {result1}")
    logger.info(f"Second square result: {result2}")

    total = result1 + result2
    logger.info(f"Total: {total}")
    return total

@app.task
async def calculate_area(length: int, width: int) -> dict:
    """
    Calculate area and perimeter using subtasks.

    Returns both area (via multiply subtask) and perimeter.
    """
    logger.info(f"Calculating area and perimeter for {length}x{width}")

    # Use multiply subtask for area
    area = await multiply(length, width)
    logger.info(f"Area calculated: {area}")

    # Calculate perimeter directly
    perimeter = 2 * (length + width)
    logger.info(f"Perimeter: {perimeter}")

    return {
        "area": area,
        "perimeter": perimeter,
        "dimensions": {"length": length, "width": width}
    }

File: trigger-tasks/src/subtasks.ts
import { task } from "@trigger.dev/sdk";
import { square, multiply } from "./basic";

export const addSquares = task({
  id: "add_squares",
  run: async (payload: { a: number; b: number }) => {
    console.log(`Computing add_squares: ${payload.a}² + ${payload.b}²`);

    const result1 = await square.triggerAndWait({ a: payload.a }).unwrap();
    console.log(`First square result: ${result1}`);

    const result2 = await square.triggerAndWait({ a: payload.b }).unwrap();
    console.log(`Second square result: ${result2}`);

    const total = result1 + result2;
    console.log(`Total: ${total}`);
    return total;
  },
});

export const calculateArea = task({
  id: "calculate_area",
  run: async (payload: { length: number; width: number }) => {
    console.log(
      `Calculating area and perimeter for ${payload.length}x${payload.width}`
    );

    const area = await multiply
      .triggerAndWait({ a: payload.length, b: payload.width })
      .unwrap();
    console.log(`Area calculated: ${area}`);

    const perimeter = 2 * (payload.length + payload.width);
    console.log(`Perimeter: ${perimeter}`);

    return {
      area,
      perimeter,
      dimensions: { length: payload.length, width: payload.width },
    };
  },
});
